### ISSUE TITLE
Incorrect Bonuses on Starships

### ISSUE THREAD

**Issue by erandisvol** [foundryvtt-starfinder/foundryvtt-starfinder#1266]
**Describe the bug**
Bonuses are inaccurate on piloting checks for gunner and pilot roles. This includes the weapon attack rolls in the weapons tab of the starship sheet.

**To Reproduce**
I created a new Starfinder world with Kevolari Venture from the compendium, and a new level 6 operative PC with 6 ranks in Piloting, 16 Dexterity, and the Operative Edge class feature. I also tested this with a different PC with a different class, and a starship made from scratch.

**Expected behavior**
The PC alone has a +14 bonus to Piloting: 6 from ranks, +3 from class skill bonus, +3 from Dexterity bonus, and +2 from Operative's Edge at sixth level. The Kevolari Venture provides a +1 bonus from its M6 Thrusters.

The total bonus for gunnery checks for this PC should be +9 (or +14 if the +3 bonus from class skill investment and +2 from Operative's Edge are included, the book doesn't say explicitly that the gunnery check is a skill check). However, each of the bonuses are reduced by 1 with no explanation in the chat card (pictured below). The resulting total bonus is +8 (including the Operative's Edge, reduced by 1).

The total bonus for Piloting checks also appears to be incorrect, with the same reductions but also with additional unexplained penalties. The pilot.skills.pil.mod is listed at +13 and the Operative's Edge at +1. This adds to 14, but obviously the numbers that reached that total aren't accurate. The M6 Thrusters are adding the +1 correctly as ship.attributes.pilotingBonus.value. But there's a -1 ship.attributes.systems.engines.mod and a -1 ship.attributes.systems.powerCore.modOther which I am unsure where they are coming from. The ship is undamaged, and all systems are nominal. The resulting total bonus is +13.

I can test checks with other roles if needed.

**Screenshots**
![image](https://github.com/foundryvtt-starfinder/foundryvtt-starfinder/assets/74384720/a3ab1204-58fd-479d-8b7b-381e8818a0dc)

![image](https://github.com/foundryvtt-starfinder/foundryvtt-starfinder/assets/74384720/be104803-c537-4536-91c7-0a16f1916397)

**Technical details (please complete the following information):**
 - OS: Windows 10
 - Browser: Edge
 - Foundry Version: 306version 11 stable
 - SFRPG Version: 0.25.2

**Comment by rmorgens** [comment]
Also applies to Foundry v11.308 using Chrome browser on Win11 with SFRPG v0.25.2.

But it looks like it is somehow applying a -1 across all possible contributers for Gunnery and Pilot
Engineer Divert has my skill focus broken out from eng.mod (possibly because it overrides the Operatives Edge??), and then drops a -3 from powerCore.mod....
Encourage is pulling -3's from lifeSupport.mod and powerCore.modOther
LockOn for Computers is pulling -3s from sensors.mod and powerCore.modOther
MaintPanelAccess has a -1 to Acrobatics skill on the actor sheet, -1 from powerCore.modOther and only 1 from my OperativesEdge which should is 2 on the actor sheet (Unlike other skills, the acr.mod seems to include my OpsEdge bonus and then there is a separate entry for it)
MO Scry has the same -1 across the board
Ship has no damage....

Trying creating ship from scratch in v308 and did not resolve

Swapped to my L6 Soldier with a +6 dex mod 6 piloting ranks and showing only 3 BAB/3 pilot ranks, only 3 dex mod and 3 skill focus on a Gunnery roll....
For piloting, the mod+focus is correct, but applying a -1 pilotingBonus.value that isn't there and -3s due to engines.mod and powerCore.modOther.....

**Comment by ian612** [comment]
I tried this with a second test scenario to try to replicate the issue/narrow down exactly what's causing it.

I ran a test scenario, checked all the crew actions, and found the following discrepancies with a brand new Kevolari Venture and crew of Level 1 iconics, all from the compendiums:

Envoy Specific (may be a separate issue)
- Envoy's skill expertise flat bonus not being added correctly (1d6 adds fine, but +1 at level 5, +2 at level 9 etc. seems to be missing)

Engineer Actions (not including Patch and Hold It Together)
- Actor's Engineering bonus is 1 lower than it should be
- `ship.attributes.systems.powerCore.mod` shows a -1, even though data inspector shows it is 0

Science Officer Actions
- Engineer class's "Bypass" feature works fine
- Actor's skill/attribute bonus is 1 lower than it should be
- `ship.attributes.systems.powerCore.modOther` shows a -1, even though data inspector shows it is 0
- `ship.attributes.systems.sensors.mod` shows a -1, even though data inspector shows it is 0
Specific to "Target System"
- `ship.attributes.sensors.mod` shows a -1, even though data inspector shows it is 0

**Comment by ian612** [comment]
After some further digging I think I understand a bit better what's happening. I was getting thrown off because some of the values weren't adding up, but I think it's because there are two simultaneous issues.

Background
If I understand correctly, for every roll there's a "tree" of nodes that gets resolved before the final formula equation can be generated and rolled. For the example of the "Divert" action, there would be two nodes:

The character's Engineering skill
The ship's Power Core Modifier

On the character sheet, the character's Engineering skill is displayed as the following (with an example value from a mechanic character I have in my test world)

skill ranks (7)
+ class skill bonus (3)
+ attribute modifier (4)
+ any applicable constant modifiers [Mechanic feature Bypass] (2)
= 16

In the roll code when the Engineering skill is calculated, it looks like it's essentially removing the value of the modifiers from the base value and then re-adding them to the roll but as separate modifiers. When you just roll a skill directly off the character sheet, this works fine since 16 - 2 + 2 = 16. 
Issue
For a starship action, this code seems to result in 2 problems:

1) The modifiers are removed from the base value of BOTH the character's Engineering skill AND the ship's Power Core Modifier before being added back, but only once. For actions where there are more than 2 nodes, each one has the modifier values removed before being added back. Thus I would expect to get the following if this was the only issue:
(16 - 2 = 14) [Engineering skill]
+ (0 - 2 = -2) [Power Core Modifier]
+ 2 [Bypass]

2) There also seems to be an issue with resolving some formulas in modifiers. The modifier formula for Bypass is equal to "1 + ceil(max(0, @classes.mechanic.levels - 4) / 4) + max(0, floor(@classes.mechanic.levels / 20))". For a mechanic at level 7, this resolves to 2. On the character sheet, this works fine. When rolling the starship action, however, this looks like it's only resolving the "1" and ignoring the "+ ceil(...)" part of the equation, which results in the actual formula that's rolled (see screenshot):
(16 - 1 = 15) [Engineering skill]
+ (0 - 1 = -1) [Power Core Modifier]
+ 1 [Bypass]

![image](https://github.com/foundryvtt-starfinder/foundryvtt-starfinder/assets/34078802/19a3675c-ee5a-498a-b7ff-0a18e4b0110f)

**Comment by danimrath** [comment]
This appears to be caused by the code in rollnode.js in resolveforroll

                // TODO: find another way to do this, or find the correct spot for this as this will result in bugs.
                const constantMods = rollMods.filter(mod => mod.modifierType === SFRPGModifierType.CONSTANT);
                const modSum = constantMods.reduce((accumulator, value) => accumulator + value.max, 0);
                this.baseValue = (Number(this.baseValue) - modSum).toString();

**Comment by ian612** [comment]
This was fixed in https://github.com/foundryvtt-starfinder/foundryvtt-starfinder/pull/1297
